resources/assets/lib/components/report-form.tsx

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import { Modal } from 'components/modal';
+import { SelectOptions } from 'components/select-options';
+import { isEmpty } from 'lodash';
+import * as _ from 'lodash';
+import * as React from 'react';
+
+const bn = 'report-form';
+const maxLength = 2000;
+
+interface Props {
+  completed: boolean;
+  disabled: boolean;
+  onClose: () => void;
+  onSubmit: ({comments}: {comments: string}) => void;
+  title: string;
+  visible: boolean;
+  visibleOptions?: string[];
+}
+
+interface ReportOption {
+  id: string;
+  text: string;
+}
+
+interface State {
+  comments: string;
+  selectedReason: ReportOption;
+}
+
+export class ReportForm extends React.PureComponent<Props, State> {
+  options = [
+    { id: 'Cheating', text: osu.trans('users.report.options.cheating') },
+    { id: 'MultipleAccounts', text: osu.trans('users.report.options.multiple_accounts') },
+    { id: 'Insults', text: osu.trans('users.report.options.insults') },
+    { id: 'Spam', text: osu.trans('users.report.options.spam') },
+    { id: 'UnwantedContent', text: osu.trans('users.report.options.unwanted_content') },
+    { id: 'Nonsense', text: osu.trans('users.report.options.nonsense') },
+    { id: 'Other', text: osu.trans('users.report.options.other') },
+  ];
+
+  constructor(props: Props) {
+    super(props);
+
+    if (props.visibleOptions) {
+      this.options = _.intersectionWith(this.options, props.visibleOptions, (left, right) => left.id === right);
+    }
+
+    this.state = {
+      comments: '',
+      selectedReason: this.options[0],
+    };
+  }
+
+  render() {
+    return this.props.visible ? this.renderForm() : null;
+  }
+
+  private handleCommentsChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    this.setState({ comments: e.target.value });
+  };
+
+  private handleReasonChange = (option: ReportOption) => {
+    this.setState({ selectedReason: option });
+  };
+
+  private renderForm() {
+    const title = this.props.completed ? osu.trans('users.report.thanks') : this.props.title;
+
+    return (
+      <Modal onClose={this.props.onClose} visible={this.props.visible}>
+        <div className={bn}>
+          <div className={`${bn}__header`}>
+            <div className={`${bn}__row ${bn}__row--exclamation`}>
+              <i className='fas fa-exclamation-triangle' />
+            </div>
+
+            <div className={`${bn}__row ${bn}__row--title`} dangerouslySetInnerHTML={{ __html: `<span>${title}</span>`}} />
+          </div>
+
+          {this.renderFormContent()}
+        </div>
+      </Modal>
+    );
+  }
+
+  private renderFormContent() {
+    const reasonElements = isEmpty(this.options) ? null :
```

Comment:
Can use `this.options.length` instead.

Also can be inlined below with `{this.options.length !== 0 && (...` instead of assigning to a variable first since there's no additional logic needed